src/ray/raylet_client/raylet_client.cc

```diff
@@ -305,13 +305,16 @@ Status raylet::RayletClient::SetResource(const std::string &resource_name,
 }
 
 void raylet::RayletClient::RequestWorkerLease(
-    const TaskSpecification &resource_spec,
+    const rpc::TaskSpec &task_spec,
     const rpc::ClientCallback<rpc::RequestWorkerLeaseReply> &callback,
     const int64_t backlog_size) {
-  rpc::RequestWorkerLeaseRequest request;
-  request.mutable_resource_spec()->CopyFrom(resource_spec.GetMessage());
-  request.set_backlog_size(backlog_size);
-  grpc_client_->RequestWorkerLease(request, callback);
+  google::protobuf::Arena arena;
```

Comment:
According to the document, this seems to allocate large memory ahead of time. Isn't it the anti-pattern to create Arena for each RPC call? (i.e. doesn't it need to be a member of RayletClient?)

Reply:
It's actually is a good pattern. Arena will not release allocated block and also it won't reuse the freed memory. 
Also to answer your other question mentioned, it'll auto increase. Internally, it's chained blocks. There is a parameter that we can tune for the size of each block.

When allocate, it'll allocate blocks and in destruction time, it'll free blocks. So protobuf won't trigger any destruction (think about complicated map objects destruction which is expensive).

In the end, all sending requests should all be built on top of the arena I think. For reply, it's not necessary and we should make replies non-const so that we can release them manually and attach them to the main thread structure.

Some references here:
>Granularity
We have found in most application server use cases that an "arena-per-request" model works well. You may be tempted to divide arena use further, either to reduce heap overhead (by destroying smaller arenas more often) or to reduce perceived thread-contention issues. However, the use of more fine-grained arenas may lead to unintended message copying, as we describe above. We have also spent effort to optimize the Arena implementation for the multithreaded use-case, so a single arena should be appropriate for use throughout a request lifetime even if multiple threads process that request.

Reply:
I see. So in this case, the deallocation will be faster (since it is just freeing blocks), but we still need to allocate memory (a list of blocks) for each request creation time, am I correct?

Reply:
I think allocation should be faster too since we can do a single allocation (and there's less fragmentation)? 

We should probably think about the per-request overhead we're introducing, but let's say we have a 100k actor scalability envelope and we allocate 10KiB (this is probably high) per actor, that's only 1GiB of overhead. 

We could probably measure this more carefully, but back of napkin estimate says it's probably a good tradeoff for us?

Reply:
Yeah agree that degree of tradeoff is worth it. One concern I have is that we might use this more often than actor creation path, and the accumulated overhead could be very large (in that case, we need some tuning like reducing block size), and the memory usage regression is really hard to detect & debug, so it’d be still great to understand the accurate overhead? Wdyt

Reply:
Maybe if we can calcuate that, there’s no need to do actual benchmark

Reply:
@rkooo567  @wuisawesome  I think it depends. First of all, it's allocating by blocks, and each block is something configurable. The next thing is that sometimes it'll use less memory since for this model, it's able to reference things, not in the request.
You can check the implementation of GetAllNodeInfo.

```
reply->add_node_info_list()->CopyFrom(*entry.second);
```
got changed to
```
    reply->mutable_node_info_list()->UnsafeArenaAddAllocated(entry.second.get());
```

In this way, we avoid copying/constructing/deconstructing, so it benefits both memory and CPU actually.

The default max block size is 8k (https://github.com/protocolbuffers/protobuf/blob/master/src/google/protobuf/arena.h#L131-L169) and the waste/gain depends on the actual request.

I'll measure it before making it ready to be merged.